Refactor ccall/cwrap into the js library

kripken opened the pull request:
I noticed that in some cases the optimizer cannot remove them from the emitted code, so this is not just nicer but also can be useful for code size, that is, ccall/cwrap will not exist at all unless the user asks for them.

To make this possible this also makes stuff in `EXPORTED_RUNTIME_METHODS` be fetched from the JS library. That is, if a user asks "export ccall from the runtime", they should not need to also add it to `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE`, which was the case before.

This is a breaking change, however - see the fixed tests. Before, just using `ccall` in your EM_ASM code would just work, since it was in the preamble, and the optimizer would know not to remove it. After this PR, you must explicitly ask for it, either using `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE`, or `EXPORTED_RUNTIME_METHODS` (the latter will also export them on Module). This is similar to other changes we made with no longer exporting stuff by default - those did cause some amount of inconvenience, but I think they were worth it.

juj commented on src/jsifier.js:
```diff
@@ -144,7 +158,7 @@ function JSify(data, functionsOnly) {
     }
 
-    function addFromLibrary(ident) {
+    function addFromLibrary(ident, firstCall) {
```
Why does this need this `firstCall` boolean? This looks somewhat arbitrary?

kripken replied:
I guess it looks odd, yeah. I'll make it more explicit, give addFromLibrary a parameter "export".

juj commented on src/jsifier.js:
```diff
@@ -110,6 +110,20 @@ function JSify(data, functionsOnly) {
+        ignoreIfMissing: true,
+        exported: true
+      });
+    });
```
Is this to provide backwards compatibility, or? If `ccall` goes in the library, does it not already follow the regular route that one uses `-s DEFAULT_LIBRARY_FUNCS_TO_INCLUDE=...` to force it being linked in if not otherwise referenced?

kripken replied:
Yes, this is partially for backwards compatibility, so that people can keep doing `EXTRA_EXPORTED_RUNTIME_METHODS=['ccall']`. This seems unambiguous, they do want it, and they want it exported. So this seems useful, and not just for backwards compatibility, to avoid needing two redundant flags to both include and export (this allows us to remove a comment in settings.js that "you must both include and export").

juj replied:
> So this seems useful, and not just for backwards compatibility, 

As we work towards removing items from preamble/runtime, I don't think we should advertise special mechanisms that we crafted around preamble/runtime to work as general items, but rather, would be good to phase those out as well. So I think retaining support for `EXTRA_EXPORTED_RUNTIME_METHODS=['ccall']` should be considered as backwards compatibility, if we keep it, because `ccall` is not a runtime/preamble function anymore.

> to avoid needing two redundant flags to both include and export (this allows us to remove a comment in settings.js that "you must both include and export").

I think this is the root bug that we should address. Noticed this just now as well, i.e. if I implement a function in a JS library that is not called anywhere, and I want to force it into the build to be exported for my custom outside world JS code to call, I need to add it to both `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE ` and `EXPORTED_FUNCTIONS`, and awkwardly, using both assembly name and C name mangling rules. This occurs in

https://github.com/emscripten-core/emscripten/pull/7667/files#diff-ae40e645f84e893ce5d037b5efc53be6R1239

Ideally we would have only one link+export mechanism that exports both C/C++ compiled functions, and also JS library functions. I.e. if I add in `-s EXPORTED_FUNCTIONS=['_foo']`, it should not matter whether `foo()` is implemented in C/C++ code or in JS library, and it should not matter whether `foo()` had any pre-existing references to it or not in some C/C++/JSlib code - it should always be linked in, and exported.

If above scheme is technically difficult, then second favorite approach would be to have a pair `-s EXPORTED_FUNCTIONS=['_foo']` and `-s EXPORTED_JS_FUNCTIONS=['_foo']`, where first directive links in and exports C/C++ functions, and second directive links in and exports JS library functions (but punt it to the developer to need to know which one it comes from). This is not as good, but probably would do the job as well, it is not hard to reason.

But it never makes sense to have to separately say "link in X" and then "export X". I.e. saying "export X" should always imply saying "link in X", because the export activity should constitute an use of X. CurrentIy adding a JS library function in to EXPORTED_FUNCTIONS means "export X, but give me an error if no C/C++/JSlib code referenced that function so it wasn't linked in via 'natural' means". That does not seem like an useful abstraction.

I wonder to fix this: can we just do in emcc.py frontend something like 

```
DEFAULT_LIBRARY_FUNCS_TO_INCLUDE  += EXPORTED_FUNCTIONS
```

(modulo C-vs-asm name mangling and compiler should ignore if some JSlib symbols in DEFAULT_LIBRARY_FUNCS_TO_INCLUDE are not found, (they could have been C/C++ symbols))

or if that's the hard part, i.e. we can't distinguish between nonexisting vs C/C++ linked symbols at that point, then perhaps with `EXPORTED_JS_FUNCTIONS` we could do effectively

```
DEFAULT_LIBRARY_FUNCS_TO_INCLUDE  += unmangle(EXPORTED_JS_FUNCTIONS)
EXPORTED_FUNCTIONS += EXPORTED_JS_FUNCTIONS
```

to build a command line directive that does both link + export a symbol?

kripken replied:
> I wonder to fix this: can we just do in emcc.py frontend something like DEFAULT_LIBRARY_FUNCS_TO_INCLUDE  += EXPORTED_FUNCTIONS

Almost, yeah - that's what this PR does for `EXPORTED_RUNTIME_METHODS`, essentially `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE += EXPORTED_RUNTIME_METHODS`.

The tricky part is to not issue an error if those things don't exist - in your example, an EXPORTED_FUNCTION may have been in compiled code, and in this PR, it may be preamble.

juj commented on tests/test_core.py:
```diff
@@ -7207,6 +7207,7 @@ def test_async(self, emterpretify=False):
 
     if emterpretify:
       print('check bad ccall use')
+      self.set_setting('EXTRA_EXPORTED_RUNTIME_METHODS', ["ccall"])
```
Given that `ccall` and `cwrap` are now regular JS library functions, wouldn't it be appropriate to use the regular JS library function export mechanism (`-s DEFAULT_LIBRARY_FUNCS_TO_INCLUDE`(?)) to export them? Using the special mechanism `EXTRA_EXPORTED_RUNTIME_METHODS` (which iiuc was crafted just for the preamble functions) to export `ccall` after it has been "demoted" from being a special preamble function to a regular JS library function looks like an odd direction?

kripken replied:
Does the above response address this?

juj replied:
I think we should wither `EXTRA_EXPORTED_RUNTIME_METHODS` away as we move functions away from preamble/runtime, since the meaning of `EXTRA_EXPORTED_RUNTIME_METHODS` was crafted specifically to cater to the special symbols found in preamble/runtime. As these symbols convert from special to regular JS library symbols, it would be good to enforce their linkage to be treated like other regular JS library symbols should be.

As backwards compatibility for `ccall` specifically and other existing symbols that were part of runtime/preamble, if we want to keep supporting the usage, we could, but I'd rather we enforced that users didn't.

Perhaps we should build a new machinery to link+export JS library functions, like described in the other comment above? It feels like perhaps we have been missing a construct here. Would that solve the issue cleanly for all JS library items?

juj commented:
This is great! +1, some questions inline.

kripken commented:
* Simplified the addFromLibrary stuff as requested.
 * Added more comments about `EXTRA_EXPORTED_RUNTIME_METHODS` vs `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE`.
 * Also added an `EXTRA_LIBRARY_FUNCS_TO_INCLUDE`, which like `EXTRA_EXPORTED_RUNTIME_METHODS` lets users add to the list, without trampling over the defaults.
 * Added useful errors in ASSERTIONS mode (see `guardAgainstUnincludedFunction`) - if a user calls call but did not include it from the JS library, we show an error saying what they need to do. This is similar to what we did when we stopped exporting stuff.

I think as we move more methods from preamble to the JS library it will be useful to have the features added here, namely
 * `EXTRA_EXPORTED_RUNTIME_METHODS` both includes and exports.
 * `EXTRA_LIBRARY_FUNCS_TO_INCLUDE` makes it easy to include more library functions.

juj commented:
I think ideally we would move towards deprecating all of the following:

```
EXPORTED_RUNTIME_METHODS  // deprecated as runtime/preamble withers away
EXTRA_EXPORTED_RUNTIME_METHODS // deprecated as runtime/preamble withers away
DEFAULT_LIBRARY_FUNCS_TO_INCLUDE // deprecated because user would use same mechanism for C/C++/JSlib symbols to link+export 
LIBRARY_DEPS_TO_AUTOEXPORT // deprecated with smarter linking tracking
ASM_PRIMITIVE_VARS // deprecated as compiler becomes smarter to track to emit only what it uses
RUNTIME_FUNCS_TO_IMPORT // deprecated as compiler becomes smarter to track to emit only what it uses
```

and we'd only leave with `EXPORTED_FUNCTIONS`, which would both link in + export a symbol, whether it comes from C/C++/JS. This would mean freezing any symbol additions to runtime/preamble (which we have been following quite well for the past year+, most of the magic variables are ancient)

(If we have to deal with the default+extra exported case, then we could have `EXPORTED_FUNCTIONS` + `EXTRA_EXPORTED_FUNCTIONS` kind of scheme, but logic being the same)

kripken commented:
@juj - agreed on the plan, yeah, would be good to end up with a single option for all exporting.

kripken commented:
It sounds like you want to do some more serious rethinking of this stuff. I can hold off on this PR in that case, it's not urgent?

juj commented on src/settings.js:
```diff
@@ -583,12 +583,15 @@ var ASYNCIFY_WHITELIST = ['qsort',
-// element, and not just methods. For example, we export the Runtime object by
+// runtime.
+// Methods on this list are automatically included from JS libraries, if they
+// are library methods.
```
Does `EXPORTED_RUNTIME_METHODS` now behave like `EXPORTED_JS_FUNCTIONS` that I was referring to in the earlier comment? I.e. if I add e.g. `-s EXPORTED_RUNTIME_METHODS=['emscripten_get_device_pixel_ratio']` and there's no other C/C++/JSlib code that would reference `emscripten_get_device_pixel_ratio()`, will the function be linked in and exported to the build?

If so, then that'd actually be great, though it does not feel good to advertise this feature on `-s EXPORTED_RUNTIME_METHODS`, but rather just silently/unadvertisingly do this to support `ccall` for backwards compatibility? And then add a new `EXPORTED_JS_FUNCTIONS` that does this same, and advertise and document it instead? That way we can move a little bit towards getting rid of the "runtime methods" concept altogether?

(sorry if I misunderstood something in this comment?)

kripken replied:
Yeah, I think your first paragraph describes how it works.

There's no special-casing of ccall here? Maybe I am misreading the second comment?

In any case, I agree with your general point - we should simplify and unify this stuff. I think it may be best to close this PR, as I don't yet have a clear picture in my head of how to make proper progress here, and the immediate ccall code size problem has been fixed in that other PR.

juj commented on src/parseTools.js:
```diff
@@ -1495,6 +1495,13 @@ function makeRetainedCompilerSettings() {
 
+function guardAgainstUnincludedFunction(name) {
+  if (!addedLibraryItems[name]) {
+    return 'function ' + name + '() { abort("ccall is now in the JS library. to use it, build with -s EXTRA_LIBRARY_FUNCS_TO_INCLUDE=[' + "'$ccall'" + ']") }\n';
```
Should this be `abort(name + " is now in the JS library ...`? This checks for `name`, but then issues a warning about `ccall`?

The name of this function feels ambiguous, does `guardAgainstUnincludedFunction` meant `guardAgainstUnincludedJsLibraryFunction`? Or `guardAgainstUnincludedPreambleFunction?`

floooh commented:
...hi, not sure if this is the right place, but I'm a bit concerned about the increased reliance on the "EXPORTED_..." linker arguments (especially from the point-of-view of a library writer) because it means the configuration of my header-only libraries need to "leak" into the library user's build-system. So instead of just "include-and-go" I need to document somewhere that the user needs to add a specific linker option.

For embedding JS code into C, EMSCRIPTEN_KEEPALIVE, EM_JS() and EM_ASM() macros are perfect solutions for this type of problem. Everything is declared in the C source, I can distribute my libs as a single C header (and not, for instance as separate C and JS source files).

What I would *love* is something similar for JS runtime functions. Let's say my library's embedded JS functions need to use 'ccall()'. It would be great if I could tell emscripten right in the source code that it needs to preserve the ccall() method instead of doing this in the linker step.

Basically something like EMSCRIPTEN_KEEPALIVE, but for JS runtime functions.

A similar, but less important feature could be to do the same for required link-libraries, that emscripten automatically resolves the required libraries (like GL) was a bit weird because it happened automatically under the hood, but for instance the Visual Studio compiler allows to define linker dependencies in the source code via the (also weirdly named) ```#pragma comment(lib, "libname")```, this is convenient for libs for the same reasons detailed above. Instead of writing in the docs "you need to link with this and this system library", the header directly contains this declaration. But as I said, this is less important than some sort of solution for the EXPORT_FUNCTIONS linker option (since libraries are much less granular than JS runtime functions)

Cheers!
-Floh.

curiousdannii commented:
If `ccall` becomes a regular library function, then your library function should be able to specify `function__deps: [ 'ccall' ]` and it should just work.

floooh commented:
The advantage of EM_JS() is that I can bundle everything into the C source, which enables STB-style cross-platform headers.

For instance here's the WebAudio backend for emscripten in sokol_audio.h:

https://github.com/floooh/sokol/blob/75a16e2ace39a5917a6296960f106d1223a9219a/sokol_audio.h#L1159-L1293

It's actually a bit more JS than C, yet I can simply include the C header and it "just works", same for other platforms. The user doesn't need to be concerned with Javascript files, and how those would be included in the C/C++ build system (or webpage).

I think EM_JS() is really one of the best features in emscripten, at least for writing cross-platform libraries. It's much, much better than what the Android NDK has to offer for C/Java interop, where C and Java needs to live in separate files, must be compiled separately etc etc...

kripken commented:
> Basically something like EMSCRIPTEN_KEEPALIVE, but for JS runtime functions.

I think we could do that with something like
```
#define EMSCRIPTEN_KEEPALIVE_JS(x)
EMSCRIPTEN_KEEPALIVE void* emscripten_keepalive_js_#x() {
  return (void*)&x;
}
```
That would keep `emscripten_keepalive_js_x` alive, which would force `x` to be included and kept alive.

I'm not sure of the value, though - js library methods are included on demand, so if someone links in a library of yours, and your library needs it, it would get included automatically. In other words, I don't see why you'd need users of your library to use a commandline flag?

floooh commented:
> In other words, I don't see why you'd need users of your library to use a commandline flag?

...let's say I have a C header with an embedded EM_JS function which needs to do a ccall, for instance here's an example from a simple asynchronous file loader:

```c
EM_JS(void, emsc_load_file, (const char* path_cstr), {
    var path = UTF8ToString(path_cstr);
    var req = new XMLHttpRequest();
    req.open("GET", path);
    req.responseType = "arraybuffer";
    req.onload = function(e) {
        var uint8Array = new Uint8Array(req.response);
        var res = ccall('emsc_load_data',
            'int',
            ['string', 'array', 'number'],
            [path, uint8Array, uint8Array.length]);
    };
    req.send();
});
```

...wouldn't I need to use a ```-s EXPORTED_RUNTIME_FUNCTIONS=['ccall']``` so that ccall() isn't dropped (at least I'm pretty sure I had to do this at one time, because it's still in my standard-linker arguments).

What I'd like to do is this, so that the runtime function requirement for ```ccall``` is close to its usage.

```c
EMSCRIPTEN_KEEPALIVE_JS(ccall);
EM_JS(void, emsc_load_file, (const char* path_cstr), {
    var path = UTF8ToString(path_cstr);
    ...
    ccall(...)
});
```

I'll try your suggestion, and report back :)

floooh commented:
Uh, ok, please forgive the noise :)

It looks like ccall is indeed automatically included for the case I described above, also without ```-s EXPORTED_RUNTIME_FUNCTIONS``` and it also works with closure enabled. Now I wonder why I ever had to add that EXPORT_RUNTIME_FUNCTIONS linker args though (until fairly recently I used Module.ccall() though, with my recent fixes for closure I changed that to a simple ```ccall()``` for code going through closure, and Module['ccall'] elsewhere (in the index.html).

I haven't tested with the MINIMAL_RUNTIME yet, but it seems like everything's good :)

kripken commented:
Thanks, now I see what you mean. So there are two sides here. On the one side, right now that will work, as ccall is included, and optimized out when not used - but in your code, the optimizer sees its used. So you only need to export it if you want to access it from outside (exported on Module).

But on the other side, this PR would have changed that - you would have had to request it even for usage like yours. I think that's another reason to not move forward with this PR, at least not without an option like you suggested like `EMSCRIPTEN_KEEPALIVE_JS`, but even that is maybe not great. Another option is that I think this may be a breaking change in MINIMAL_RUNTIME anyhow, so we'll need to figure things out properly there eventually.

floooh commented:
Ok, I'll see if I can play around with the MINIMAL_RUNTIME stuff a bit more in the next days and see how it goes. Thanks for the quick responses :)

floooh commented:
...I think I have a solution for exporting things like ccall/cwrap to be callable from the "outside" (like the .html file) despite minification:

Have a EM_JS() function which registers the required emscripten runtime functions on Module, and call this function from the C code during application init, for instance:

```cpp
EM_JS(void, emsc_fs_init, (void), {
    Module['ccall'] = ccall;
});
```
Each of my library headers knows what functions should be callable from the outside, and add them accordingly. I think this is a good enough solution for my use case.

sbc100 commented:
The deletion of the `incoming` branch triggered the automatic closing of this branch.  My apologies.  This was an unexpected side effect.

Re-opening and re-targeting to master.

sbc100 commented:
bump

sbc100 commented:
I think this was superseded by #17433 and can be closed?

@kripken if you agree can you close this and delete the branch?

kripken commented:
Agreed.
